File: go/replicant/client_manager.go
package replicant

import (
	"github.com/psu-csl/replicated-store/go/multipaxos"
	logger "github.com/sirupsen/logrus"
	"net"
	"sync"
)

type ClientManager struct {
	nextId       int64
	clientIdMu   sync.Mutex
	numPeers     int64
	multipaxos   *multipaxos.Multipaxos
	mu           sync.Mutex
	clients      map[int64]*Client
	isFromClient bool
}

func NewClientManager(id int64,
	numPeers int64,
	mp *multipaxos.Multipaxos,
	isFromClient bool) *ClientManager {
	cm := &ClientManager{
		nextId:       id,
		numPeers:     numPeers,
		multipaxos:   mp,
		clients:      make(map[int64]*Client),
		isFromClient: isFromClient,
	}
	return cm
}

func (cm *ClientManager) NextClientId() int64 {
	cm.clientIdMu.Lock()
	defer cm.clientIdMu.Unlock()
	id := cm.nextId
	cm.nextId += cm.numPeers
	return id
}

func (cm *ClientManager) Start(socket net.Conn) {
	id := cm.NextClientId()
	client := NewClient(id, socket, cm.multipaxos, cm, cm.isFromClient)

	cm.mu.Lock()
	cm.clients[id] = client
	cm.mu.Unlock()
	logger.Infof("client_manager started client %v\n", id)
	go client.Start()
}

func (cm *ClientManager) Get(id int64) *Client {
	cm.mu.Lock()
	defer cm.mu.Unlock()
	client, ok := cm.clients[id]
	if !ok {
		return nil
	}
	return client
}

func (cm *ClientManager) Stop(id int64) {
	cm.mu.Lock()
	defer cm.mu.Unlock()
	logger.Infof("client_manager stopped client %v\n", id)
	client, ok := cm.clients[id]
	if !ok {
		panic("no client to stop")
	}
	client.Stop()
	delete(cm.clients, id)
}

func (cm *ClientManager) StopAll() {
	cm.mu.Lock()
	defer cm.mu.Unlock()
	for id, client := range cm.clients {
		logger.Infof("client_manager stopping all clients %v\n", id)
		client.Stop()
		delete(cm.clients, id)
	}
}

File: c++-tcp/client_manager.cc
#include "client_manager.h"
#include <glog/logging.h>

using asio::ip::tcp;

void ClientManager::Start(tcp::socket socket) {
  auto id = NextClientId();
  auto client =
    std::make_shared<Client>(id, std::move(socket), multi_paxos_, this,
                             is_from_client_);
  {
    std::scoped_lock lock(mu_);
    auto [it, ok] = clients_.insert({id, client});
    CHECK(ok);
  }
  DLOG(INFO) << " client_manager started client " << id;
  client->Start();
}

client_ptr ClientManager::Get(int64_t id) {
  std::scoped_lock lock(mu_);
  auto it = clients_.find(id);
  if (it == clients_.end())
    return nullptr;
  return it->second;
}

void ClientManager::Stop(int64_t id) {
  std::scoped_lock lock(mu_);
  DLOG(INFO) << " client_manager stopped client " << id;
  auto it = clients_.find(id);
  CHECK(it != clients_.end());
  it->second->Stop();
  clients_.erase(it);
}

void ClientManager::StopAll() {
  std::scoped_lock lock(mu_);
  for (auto& [id, client] : clients_) {
    DLOG(INFO) << " client_manager stopping all clients " << id;
    client->Stop();
  }
  clients_.clear();
  //thread_pool_.join();
}
